Upstream redesigned text cursor support for WebKit

rr-codes opened the pull request:
#### f9f5d07097d5ca191d4f7241c04e6b1b684c8352
<pre>
Upstream redesigned text cursor support for WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=257887">https://bugs.webkit.org/show_bug.cgi?id=257887</a>
rdar://110483283

Reviewed by Mike Wyrzykowski and Tim Horton.

Upstream support for the redesigned text cursor in WebKit.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::createCaretAnimator):
(WebCore::FrameSelection::caretNode):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/platform/CaretAnimator.h:
(WebCore::currentTimeSinceEpoch):
(WebCore::platformCaretColor):
* Source/WebCore/platform/DictationCaretAnimator.cpp: Added.
(WebCore::keyframe):
(WebCore::tailBlurRadius):
(WebCore::caretBlurRadius):
(WebCore::coneStart):
(WebCore::coneEnd):
(WebCore::DictationCaretAnimator::keyframeCount const):
(WebCore::DictationCaretAnimator::DictationCaretAnimator):
(WebCore::DictationCaretAnimator::keyframeTimeDelta const):
(WebCore::DictationCaretAnimator::setBlinkingSuspended):
(WebCore::DictationCaretAnimator::computeTailRect const):
(WebCore::DictationCaretAnimator::computeScrollLeft const):
(WebCore::DictationCaretAnimator::updateGlowTail):
(WebCore::DictationCaretAnimator::updateAnimationProperties):
(WebCore::DictationCaretAnimator::start):
(WebCore::DictationCaretAnimator::debugDescription const):
(WebCore::DictationCaretAnimator::resetGlowTail):
(WebCore::DictationCaretAnimator::tailRect const):
(WebCore::DictationCaretAnimator::makeDictationTailConePath const):
(WebCore::DictationCaretAnimator::fillCaretTail const):
(WebCore::DictationCaretAnimator::expandedCaretRect const):
(WebCore::DictationCaretAnimator::paint const):
(WebCore::DictationCaretAnimator::isLeftToRightLayout const):
(WebCore::DictationCaretAnimator::caretRepaintRectForLocalRect const):
(WebCore::DictationCaretAnimator::stop):
* Source/WebCore/platform/DictationCaretAnimator.h: Added.
* Source/WebCore/platform/OpacityCaretAnimator.cpp: Added.
(WebCore::OpacityCaretAnimator::OpacityCaretAnimator):
(WebCore::OpacityCaretAnimator::keyframeTimeDelta const):
(WebCore::OpacityCaretAnimator::setBlinkingSuspended):
(WebCore::OpacityCaretAnimator::updateAnimationProperties):
(WebCore::OpacityCaretAnimator::start):
(WebCore::OpacityCaretAnimator::debugDescription const):
(WebCore::OpacityCaretAnimator::paint const):
(WebCore::OpacityCaretAnimator::caretRepaintRectForLocalRect const):
* Source/WebCore/platform/OpacityCaretAnimator.h: Added.
* Source/WebCore/rendering/CaretRectComputation.cpp:
(WebCore::caretWidth):
* Source/WebCore/rendering/CaretRectComputation.h:
(WebCore::redesignedTextCursorEnabled):
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::canShowCapsLockIndicator):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::radiiForUnderline):
(WebCore::trimRadii):
(WebCore::snapRectToDevicePixelsInDirection):
(WebCore::layoutBoxSequenceLocation):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::fillCompositionUnderline const):
* Source/WebKit/Platform/cocoa/_WKWebViewTextInputNotifications.h: Added.
* Source/WebKit/Platform/cocoa/_WKWebViewTextInputNotifications.mm: Added.
(-[_WKWebViewTextInputNotifications caretType]):
(-[_WKWebViewTextInputNotifications initWithWebView:]):
(-[_WKWebViewTextInputNotifications dictationDidStart]):
(-[_WKWebViewTextInputNotifications dictationDidEnd]):
(-[_WKWebViewTextInputNotifications dictationDidPause]):
(-[_WKWebViewTextInputNotifications dictationDidResume]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::updateCaretDecorationPlacement):
(WebKit::subscribeToTextInputNotifications):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/265388@main">https://commits.webkit.org/265388@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53fc6ccd86bddf99ed617041d68e6bfeeb4e7c27

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10685 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12334 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10244 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10700 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13279 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13156 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11758 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8978 "7 api tests failed or timed out") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12736 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9052 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9637 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16895 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10122 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9788 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13042 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10256 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8349 "1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9416 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13688 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1212 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10118 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -58,6 +58,8 @@ class CaretAnimationClient {
 
     virtual Document* document() = 0;
+
+    virtual Node* caretNode() = 0;
```
previously, DictationCaretAnimator was accessing FrameSelection, and since FrameSelection includes DictationCaretAnimator, that seemed a bit wrong. So I introduced this method on CaretAnimationClient to remove that "cycle"

mwyrzykowski replied:
Normally I wouldn't suggest fixing this during upstreaming but as this change is isolated and a good one, I think we can leave it included in this PR.

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6070,16 +6087,29 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
-#import <WebKitAdditions/WebViewImplAdditions.mm>
-#else
+#if HAVE(REDESIGNED_TEXT_CURSOR)
 void WebViewImpl::updateCaretDecorationPlacement()
```
note to self: rename to `cursor accessory `

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6070,16 +6087,29 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
+    if ([textInputNotifications caretType] == WebCore::CaretAnimatorType::Alternate) {
+        // The dictation cursor accessory should always be visible no matter what, since it is
+        // the only prominent way a user can tell if dictation is active.
+        context.showsCursorAccessories = YES;
```
note to self: add SPI declaration

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -945,7 +947,9 @@ ALLOW_DEPRECATED_DECLARATIONS_END
 #endif
 
+#if HAVE(REDESIGNED_TEXT_CURSOR)
     RetainPtr<NSObject> _textInputNotifications;
```
should this be changed to _WKWebViewTextInputNotifications instead of NSObject?

mwyrzykowski replied:
If you did, you could remove the dynamic downcast in ` WebViewImpl::updateCursorAccessoryPlacement`

Probably outside the scope of this PR, but up to you

hortont424 commented on Source/WebCore/platform/DictationCaretAnimator.cpp:
```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (C) 2004-2022 Apple Inc. All rights reserved.
```
A surprising date range

mwyrzykowski reviewed (changes requested):
We should wait until rdar://problem/110235298 merges prior to upstreaming, otherwise we will lose those changes

pxlcoder commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1922,16 +1950,35 @@ void CaretBase::paintCaret(const Node& node, GraphicsContext& context, const Lay
+#endif
+
+            return oldCaretColor;
+        }();
```
Shouldn't all this logic be in `CaretBase::computeCaretColor`? Seems strange to "compute" the color again in the painting method?

rr-codes replied:
yes; I planned on doing that refactor in a separate PR; didn't want to refactor any more than necessary in this one

mwyrzykowski replied:
Why is this change even necessary as part of the upstreaming? I think there should be as few lines changed as possible during upstreaming to avoid potential regressions. Is it wrong to have:
```
if (caretAnimator)
        caretAnimator->paint(node, context, pixelSnappedCaretRect, caretColor, paintOffset, selection);
    else
        context.fillRect(pixelSnappedCaretRect, caretColor);
```
as we did before?

rr-codes replied:
@mwyrzykowski I originally had this, but ended up changing it because `platformCaretColor` was in one of the caret animators and it compiled fine previously, but would not compile after upstreaming the animator files, so I had to put it elsewhere and FrameSelection was a logical place

mwyrzykowski replied:
Wouldn't as a static member function on the base class `CaretAnimator` be more appropriate since `platformCaretColor` is being using by `OpacityCaretAnimator` and `DictationCaretAnimator`?

rr-codes replied:
yes, that makes more sense. will change for this PR, and then combine platformcaretcolor with computecaretcolor in a follow up PR.

mwyrzykowski commented on Source/WebKit/Platform/cocoa/_WKWebViewTextInputNotifications.h:
```diff
@@ -0,0 +1,46 @@
+
+@interface _WKWebViewTextInputNotifications : NSObject
+- (instancetype)initWithWebView:(WebKit::WebViewImpl*)webView;
+- (void)dictationDidStart;
```
`dicitationDid*` doesn't need to be in the header. Probably outside the scope of this PR

mwyrzykowski commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -274,6 +274,8 @@ class FrameSelection final : private CaretBase, public CaretAnimationClient {
     WEBCORE_EXPORT void caretAnimatorInvalidated(CaretAnimatorType);
 #endif
 private:
+    static Color platformCaretColor(const RenderStyle& elementStyle, const Node*);
```
Why is this part of `FrameSelection`? I don't see `::platformCaretColor` anywhere? Only `static Color platformCaretColor(const RenderStyle& elementStyle, const Node* node)` which is a static function with a different signature than this member function

rr-codes replied:
that is... a very good point. how did this even compile

mwyrzykowski replied:
Presumably no one uses it so there is no linker error

mwyrzykowski commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -90,7 +92,7 @@ class CaretAnimator {
     virtual void setVisible(bool) = 0;
 
     PresentationProperties presentationProperties() const { return m_presentationProperties; }
-    virtual void paint(const Node&, GraphicsContext&, const FloatRect&, const Color&, const LayoutPoint&, const std::optional<VisibleSelection>&) const;
```
why is this change being made during upstreaming?

rr-codes replied:
good point. will undo

rr-codes committed:
Upstream redesigned text cursor support for WebKit
https://bugs.webkit.org/show_bug.cgi?id=257887
rdar://110483283

Reviewed by Mike Wyrzykowski and Tim Horton.

Upstream support for the redesigned text cursor in WebKit.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::createCaretAnimator):
(WebCore::FrameSelection::caretNode):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/platform/CaretAnimator.h:
(WebCore::currentTimeSinceEpoch):
(WebCore::platformCaretColor):
* Source/WebCore/platform/DictationCaretAnimator.cpp: Added.
(WebCore::keyframe):
(WebCore::tailBlurRadius):
(WebCore::caretBlurRadius):
(WebCore::coneStart):
(WebCore::coneEnd):
(WebCore::DictationCaretAnimator::keyframeCount const):
(WebCore::DictationCaretAnimator::DictationCaretAnimator):
(WebCore::DictationCaretAnimator::keyframeTimeDelta const):
(WebCore::DictationCaretAnimator::setBlinkingSuspended):
(WebCore::DictationCaretAnimator::computeTailRect const):
(WebCore::DictationCaretAnimator::computeScrollLeft const):
(WebCore::DictationCaretAnimator::updateGlowTail):
(WebCore::DictationCaretAnimator::updateAnimationProperties):
(WebCore::DictationCaretAnimator::start):
(WebCore::DictationCaretAnimator::debugDescription const):
(WebCore::DictationCaretAnimator::resetGlowTail):
(WebCore::DictationCaretAnimator::tailRect const):
(WebCore::DictationCaretAnimator::makeDictationTailConePath const):
(WebCore::DictationCaretAnimator::fillCaretTail const):
(WebCore::DictationCaretAnimator::expandedCaretRect const):
(WebCore::DictationCaretAnimator::paint const):
(WebCore::DictationCaretAnimator::isLeftToRightLayout const):
(WebCore::DictationCaretAnimator::caretRepaintRectForLocalRect const):
(WebCore::DictationCaretAnimator::stop):
* Source/WebCore/platform/DictationCaretAnimator.h: Added.
* Source/WebCore/platform/OpacityCaretAnimator.cpp: Added.
(WebCore::OpacityCaretAnimator::OpacityCaretAnimator):
(WebCore::OpacityCaretAnimator::keyframeTimeDelta const):
(WebCore::OpacityCaretAnimator::setBlinkingSuspended):
(WebCore::OpacityCaretAnimator::updateAnimationProperties):
(WebCore::OpacityCaretAnimator::start):
(WebCore::OpacityCaretAnimator::debugDescription const):
(WebCore::OpacityCaretAnimator::paint const):
(WebCore::OpacityCaretAnimator::caretRepaintRectForLocalRect const):
* Source/WebCore/platform/OpacityCaretAnimator.h: Added.
* Source/WebCore/rendering/CaretRectComputation.cpp:
(WebCore::caretWidth):
* Source/WebCore/rendering/CaretRectComputation.h:
(WebCore::redesignedTextCursorEnabled):
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::canShowCapsLockIndicator):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::radiiForUnderline):
(WebCore::trimRadii):
(WebCore::snapRectToDevicePixelsInDirection):
(WebCore::layoutBoxSequenceLocation):
(WebCore::TextBoxPainter<TextBoxPath>::fillCompositionUnderline const):
* Source/WebKit/Platform/cocoa/_WKWebViewTextInputNotifications.h: Added.
* Source/WebKit/Platform/cocoa/_WKWebViewTextInputNotifications.mm: Added.
(-[_WKWebViewTextInputNotifications caretType]):
(-[_WKWebViewTextInputNotifications initWithWebView:]):
(-[_WKWebViewTextInputNotifications dictationDidStart]):
(-[_WKWebViewTextInputNotifications dictationDidEnd]):
(-[_WKWebViewTextInputNotifications dictationDidPause]):
(-[_WKWebViewTextInputNotifications dictationDidResume]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::updateCaretDecorationPlacement):
(WebKit::subscribeToTextInputNotifications):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/265388@main
